File: sem3/graph/1b.cpp
#include <bits/stdc++.h>
 
using namespace std;

vector<map<long long, long long>> g;
vector<long long> p;
vector<bool> v;

void dfs(long long i) {
    v[i] = true;
    for (auto j : g[i]) {
        if (!v[j.first]) {
            dfs(j.first);
        }
    }
    p.push_back(i);
}

int main() {
    long long n, m, s, t;
    cin >> n >> m >> s >> t;
    --s; --t;
    g.resize(n);
    
    for (long long i = 0; i < m; ++i) {
        long long a, b, c;
        cin >> a >> b >> c;
        if (g[a - 1].count(b - 1) == 0) {
            g[a - 1][b - 1] = c;
        } else {
            g[a - 1][b - 1] = min(g[a - 1][b - 1], c);
        }
    }

    v.assign(n, false);
    dfs(s);

    vector<long long> r(n, 1e18);
    r[s] = 0;
    for (long long i = p.size() - 1; i >= 0; --i) {
        if (r[p[i]] == 1e18) {
            continue;
        }
        for (auto j : g[p[i]]) {
            r[j.first] = min(r[j.first], r[p[i]] + j.second);
        }
    }


    if (r[t] == 1e18) {
        cout << "Unreachable";
    } else {
        cout << r[t];
    }
}

File: sem3/graph/1b.py
import sys
sys.setrecursionlimit(150000)

n, m, s, t = map(int, input().split())
s -= 1
t -= 1
g = [dict() for i in range(n)]

for _ in range(m):
    a, b, c = map(int, input().split())
    if not (b - 1 in g[a - 1]):
        g[a - 1][b - 1] = c
    else:
        g[a - 1][b - 1] = min(g[a - 1][b - 1], c)

v = [False] * n
p = []
def dfs(i):
    v[i] = True
    for j in g[i]:
        if not v[j]:
            dfs(j)
    p.append(i)
dfs(s)
p.reverse()

r = [999999999] * n
r[s] = 0
for i in p:
    if r[i] == 999999999:
        continue
    for j in g[i]:
        r[j] = min(r[j], r[i] + g[i][j])
if r[t] == 999999999:
    print('Unreachable')
else:
    print(r[t])